apis/pkg/v1alpha1/lock.go

```diff
@@ -0,0 +1,124 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/crossplane/crossplane/pkg/dag"
+)
+
+var _ dag.Node = &Dependency{}
+var _ dag.Node = &LockPackage{}
+
+// A PackageType is a type of package.
+type PackageType string
+
+// Types of packages.
+const (
+	ConfigurationPackageType PackageType = "Configuration"
+	ProviderPackageType      PackageType = "Provider"
+)
+
+// LockPackage is a package that is in the lock.
+type LockPackage struct {
+	Name         string       `json:"name"`
+	Type         PackageType  `json:"type"`
+	Source       string       `json:"source"`
+	Version      string       `json:"version"`
+	Dependencies []Dependency `json:"dependencies"`
+}
+
+// ToNodes converts LockPackages to DAG nodes.
+func ToNodes(pkgs []LockPackage) []dag.Node {
+	nodes := make([]dag.Node, len(pkgs))
+	for i, r := range pkgs {
+		r := r // Pin range variable so we can take its address.
+		nodes[i] = &r
+	}
+	return nodes
+}
+
+// Identifier returns the source of a LockPackage.
+func (l *LockPackage) Identifier() string {
+	return l.Source
+}
+
+// Neighbors returns dependencies of a LockPackage.
+func (l *LockPackage) Neighbors() []dag.Node {
+	nodes := make([]dag.Node, len(l.Dependencies))
+	for i, r := range l.Dependencies {
+		r := r // Pin range variable so we can take its address.
+		nodes[i] = &r
+	}
+	return nodes
+}
+
+// AddNeighbors adds dependencies to a LockPackage.
+// NOTE(hasheddan): a LockPackage should always have all dependencies declared,
+// so we no-op when adding a neighbor.
+func (l *LockPackage) AddNeighbors(nodes ...dag.Node) error {
+	return nil
+}
+
+// A Dependency is a dependency of a package in the lock.
+type Dependency struct {
+	Package     string      `json:"package"`
+	Type        PackageType `json:"type"`
+	Constraints string      `json:"constraints"`
+}
+
+// Identifier returns a dependency's source.
+func (d *Dependency) Identifier() string {
+	return d.Package
+}
+
+// Neighbors in is a no-op for dependencies because we are not yet aware of its
+// dependencies.
+func (d *Dependency) Neighbors() []dag.Node {
+	return nil
+}
+
+// AddNeighbors is a no-op for dependencies. We should never be adding neighbors
+// to a dependency.
+func (d *Dependency) AddNeighbors(...dag.Node) error {
+	return nil
+}
+
+// +kubebuilder:object:root=true
+// +genclient
+// +genclient:nonNamespaced
+
+// Lock is the CRD type that tracks package dependencies.
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
+// +kubebuilder:resource:scope=Cluster
+type Lock struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Packages []LockPackage `json:"packages,omitempty"`
```

Comment:
I think it would be somewhat of a violation of its purpose if we did. I would prefer that we reflect any information as events on revisions or conditions in their status.